### PR TITLE
OoS for SE

### PR DESCRIPTION
**Pull request by Feretj**
#### What does this implement/fix? Explain your changes.
It implement out-of-sample extension for Spectral Embedding with knn kernel. For that it also make SE to save training data and diag from laplacian. 

#### Any other comments?
Algorithm is token from article "Out-of-Sample Extensions for LLE, Isomap,
MDS, Eigenmaps, and Spectral Clustering" Yoshua Bengio, Jean-Franc¸ois Paiement and Pascal Vincent.

### CODE REVIEW

File: sklearn/manifold/spectral_embedding_.py [hunk 1]
```diff
@@ -494,10 +501,12 @@ def fit(self, X, y=None):
                               "name or a callable. Got: %s") % self.affinity)
 
         affinity_matrix = self._get_affinity_matrix(X)
-        self.embedding_ = spectral_embedding(affinity_matrix,
-                                             n_components=self.n_components,
-                                             eigen_solver=self.eigen_solver,
-                                             random_state=random_state)
+        (self.embedding_,
```
Comment (YuryYa) [inline sklearn/manifold/spectral_embedding_.py:504]: Why \n here?

**Review by YuryYa — APPROVED** [review]

File: sklearn/manifold/spectral_embedding_.py [hunk 2]
```diff
@@ -520,3 +529,42 @@ def fit_transform(self, X, y=None):
...
+            and n_features is the number of features.
+
+        Returns
+        -------
+        X_new : array, shape = (n_samples, n_components)
+        """
+        X = check_array(X)
+
+        if (self.affinity == "nearest_neighbors"):
+            d = pairwise_distances(X, self.training_data_, n_jobs=self.n_jobs,
```
Comment (NickVeld) [inline sklearn/manifold/spectral_embedding_.py]: The common codestyle say that lines of arguments except the last must be approximately equal

File: sklearn/manifold/spectral_embedding_.py [hunk 3]
```diff
@@ -520,3 +529,42 @@ def fit_transform(self, X, y=None):
...
+                                   metric="euclidean")
+            d[(d == 0)] = np.inf
+            M = np.zeros(d.shape)
```
Comment (NickVeld) [inline sklearn/manifold/spectral_embedding_.py]: one-letter names (for example "M", "d") are bad names for variables
(dd is also not a self-commented name)

File: sklearn/manifold/spectral_embedding_.py [hunk 4]
```diff
@@ -520,3 +529,42 @@ def fit_transform(self, X, y=None):
...
+                                 axis=0)[self.n_neighbors - 1]
+            for i in range(M.shape[1]):
+                M[(d[:, i] <= kth_d[i]), i] += 0.5
+        else:
+            raise NotImplementedError(("%s is not supported. Expected "
+                                      "'nearest_neighbors' affinity."
+                                       ) % self.affinity)
+
+        M /= self._dd * self._dd
+        return np.matmul(M, self.embedding_)
```
Comment (NickVeld) [inline sklearn/manifold/spectral_embedding_.py]: The last line must be blank

File: sklearn/manifold/spectral_embedding_.py (hunk 1 above)
Comment (Feretj) [reply sklearn/manifold/spectral_embedding_.py:504]: To ensure that the line is not too long

**Pull request merged**